src/app/gui/queryresults/queryresultsservice.js

```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
+
+    /** @FIXME add description */
+    else {
+      feature.selection.selected = !feature.selection.selected;
+    }
+
+    /** @FIXME add description */
+    this.mapService.setSelectionFeatures(
+      feature.selection.selected ? 'add' : 'remove',
+      { feature: layer.selection.features.find(selectionFeature => feature.id === selectionFeature.getId()) }
+    );
+
+    // Set selection layer active based on features selection selected properties
     layer.selection.active = layer.selection.features.reduce((accumulator, feature) => accumulator || feature.selection.selected, false)
-  } else { // case of project layer on TOC
-    const fid = feature ? feature.attributes[G3W_FID]: null;
+
+  }
+  
+  /**
+   * A project layer on TOC
+   */
+  if (false === (layer.external && typeof "undefined" !== layer.external)) {
+
+    const fid = feature ? feature.attributes[G3W_FID] : null;
     const hasAlreadySelectioned = layer.getFilterActive() || layer.hasSelectionFid(fid);
-    if (!hasAlreadySelectioned) {
-      if (feature && feature.geometry && !layer.getOlSelectionFeature(fid)) {
-        layer.addOlSelectionFeature({
-          id: fid,
-          feature
-        })
-      }
+    
+    /** @FIXME add description */
+    if (!hasAlreadySelectioned && feature && feature.geometry && !layer.getOlSelectionFeature(fid)) {
+      layer.addOlSelectionFeature({ id: fid, feature });
     }
-    if (force === undefined) layer[hasAlreadySelectioned ? 'excludeSelectionFid': 'includeSelectionFid'](fid);
-    else if (!hasAlreadySelectioned && force === 'add') await layer.includeSelectionFid(fid);
-    else if (hasAlreadySelectioned && force === 'remove') await layer.excludeSelectionFid(fid);
+    
+    /** @FIXME add description */
+    if (undefined === force) {
+      layer[hasAlreadySelectioned ? 'excludeSelectionFid': 'includeSelectionFid'](fid);
+    }
+
+    /** @FIXME add description */
+    if (undefined !== force && !hasAlreadySelectioned && 'add' === force) {
+      await layer.includeSelectionFid(fid);
+    }
+
+    /** @FIXME add description */
+    if (undefined !== force && hasAlreadySelectioned && 'remove' === force) {
+      await layer.excludeSelectionFid(fid);
+    }
+
+    /** @FIXME add description */
     if (layer.getFilterActive()) {
+
       const currentLayer = this.state.layers.find(_layer => _layer.id === layer.getId());
-      layer.getSelectionFids().size > 0 && currentLayer && currentLayer.features.splice(index, 1);
+
+      /** @FIXME add description */
+      if (layer.getSelectionFids().size > 0 && currentLayer) {
+        currentLayer.features.splice(index, 1);
+      }
+
       this.mapService.clearHighlightGeometry();
-      this.state.layers.length === 1 && !this.state.layers[0].features.length && this.state.layers.splice(0);
+
+      /** @FIXME add description */
+      if (1 === this.state.layers.length && !this.state.layers[0].features.length) {
+        this.state.layers.splice(0);
+      }
+
     }
+
   }
+
```

Comment:
within this function `QueryResultsService::_addRemoveSelectionFeaturelayer, feature, index, force)`, i rewrote several if-else conditions to reduce nesting level, check carefully